### PR TITLE
Unify auth

### PR DESCRIPTION
**Pull request by mmaney**
Contributor offers to license certain software (a “Contribution” or multiple
“Contributions”) to sewer, and sewer agrees to accept said Contributions,
under the terms of the MIT License.
Contributor understands and agrees that sewer shall have the irrevocable and perpetual right to make
and distribute copies of any Contribution, as well as to create and distribute collective works and
derivative works of any Contribution, under the MIT License.

## What(What have you changed?)

Enough of the refactoring of Auth providers that I want to get it out in the open.  High points:
* future is BaseAuth and sewer.providers; dns_providers is legacy; http_providers is toast
* very simple http_dummy provider for testing and to illustrate new interface
* client uses the batch interface (clear_cert) but still uses setup_auth to do them one at a time.  What can I say, I forgot to get back to that.

Of course there's more to do, but this felt like a good place to checkpoint.  Except I might push that change to make client use setup_cert later...

## Why(Why did you change it?)

Because it was there.  :-)

### CODE REVIEW

**Comment by komuw** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 4
           

Complexity increasing per file
==============================
- sewer/auth.py  4
- sewer/providers/http_dummy.py  2
- sewer/providers/tests/test_dummy.py  2
         

Clones removed
==============
+ sewer/auth.py  -1
+ sewer/client.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5306180&bid=17365943)

File: sewer/providers/tests/test_dummy.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+from unittest import mock
```
Comment (komuw) [inline sewer/providers/tests/test_dummy.py:1]: Codacy Issue found: [Unused mock imported from unittest](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5306180)

File: sewer/auth.py [hunk 2]
```diff
@@ -1,42 +1,116 @@
 import logging
+from typing import Sequence, Tuple
 
+from hashlib import sha256
+from sewer.lib import safe_base64
```
Comment (komuw) [inline sewer/auth.py:5]: Codacy Issue found: [Unused safe_base64 imported from sewer.lib](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5306180)

File: sewer/auth.py [hunk 3]
```diff
@@ -1,42 +1,116 @@
 import logging
+from typing import Sequence, Tuple
 
+from hashlib import sha256
```
Comment (komuw) [inline sewer/auth.py:4]: Codacy Issue found: [Unused sha256 imported from hashlib](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5306180)

File: sewer/dns_providers/common.py [hunk 4]
```diff
@@ -1,20 +1,52 @@
...
 
     return safe_base64(sha256(key_auth.encode("utf8")).digest())
 
 
-class BaseDns(BaseAuthProvider):
-    def __init__(self):
-        super(BaseDns, self).__init__("dns-01")
+class BaseDns(BaseAuth):
+    """
+    BaseDns provides a legacy compatibility layer in the *_auth methods.
```
Comment (komuw) [inline sewer/dns_providers/common.py:17]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5306180)

**Comment by komuw** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- sewer/providers/http_dummy.py  2
- sewer/providers/tests/test_dummy.py  2
- sewer/auth.py  4
         

Clones removed
==============
+ sewer/auth.py  -1
+ sewer/client.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/komuW/sewer/pullRequest?prid=5306180&bid=17365943)

**Comment by komuw** [comment]
cc @donspaulding @AlecTroemel what do you think?

File: sewer/auth.py [hunk 5]
```diff
@@ -1,42 +1,115 @@
...
-        self.logger.setLevel(self.LOG_LEVEL)
-        if not self.logger.hasHandlers():
-            handler = logging.StreamHandler()
-            formatter = logging.Formatter("%(message)s")
-            handler.setFormatter(formatter)
-            self.logger.addHandler(handler)
 
-        self.auth_type = auth_type
+supported_validations = {
+    "http-01": ValidationMethod("http-01", "dns", True),
```
Comment (komuw) [inline sewer/auth.py:12]: should `"dns", True` for this?

Comment (mmaney) [reply sewer/auth.py:12]: tl;dr: ACME has two type: fields for different things.  Stop now unless you enjoy the pain of perusing RFCs over and over.  :-)

Yes.  Though the True at the end is either an artifact of history not recorded in a commit, or an accommodation for a future change (list of all known, whether supported or not?) not yet implemented.

The "dns" is the authorization type, and is short for "domain-name", the sort sewer knows about to date.  Recently an "ip" type has been added.  I'm not certain of the use for that except for an impression that it's aimed at hosting providers and other large-scale services.

The familiar "http-01" and "dns-01" (and that new tls one whose spelling I don't care to lookup now) identify the challenge type, which might better be called the challenge protocol.  The RFCs - more than just 8555 is involved, and there's a registry - add to the confusion by calling the challenge type "label", as in rfc8555 9.7.7:

`Label: The value to be put in the "type" field of the identifier object`

There's an example of the result of querying an unsatisfied authorization somewhere in the RFC - that's why we have to sift through them to find the one that is for dns-01 (or now, http-01) to get the magic values used in forming the challenge response.

File: sewer/client.py [hunk 6]
```diff
@@ -250,18 +250,6 @@ def _request(self, method, url, **kwargs):
 
         return response
 
-    @staticmethod
```
Comment (AlecTroemel) [inline sewer/client.py:253]: maybe these changes to log_response could be in their own PR?  I feel like this and the auth changes are 2 separate concerns/discussions

File: sewer/client.py [hunk 7]
```diff
@@ -722,10 +713,8 @@ def get_certificate(self):
                 identifier_auth = self.get_identifier_authorization(url)
                 token = identifier_auth["token"]
                 acme_keyauthorization = self.get_keyauthorization(token)
-                cleanup_kwargs = self.auth_provider.fulfill_authorization(
-                    identifier_auth, token, acme_keyauthorization
-                )
-                cleanup_kwargs_list.append(cleanup_kwargs)
+                self.auth_provider.setup_auth(identifier_auth, token, acme_keyauthorization)
```
Comment (AlecTroemel) [inline sewer/client.py:716]: I'd like to see how `setup_cert` would fit in here, since its counterpart `clear_cert` is being used below

File: sewer/auth.py [hunk 8]
```diff
@@ -1,42 +1,115 @@
...
+        else:
+            self.logger = logging.getLogger(__name__)
+            self.logger.setLevel(log_level)
+            if not self.logger.hasHandlers():
+                handler = logging.StreamHandler()
+                formatter = logging.Formatter("%(message)s")
+                handler.setFormatter(formatter)
+                self.logger.addHandler(handler)
+
+    def setup_cert(self, authorizations: Sequence[Tuple[str, str, str]]):
```
Comment (AlecTroemel) [inline sewer/auth.py:92]: I'm not sure `setup_cert` and `clear_cert` are the right names for these methods, but I have nothing better right now. I'll need to give it more thought. 

also will this method need to return the `cleanup_list` used in client.py? something like this?

```python
def setup_cert(self, authorizations: Sequence[Tuple[str, str, str]]) ->  Sequence[Tuple[str, str, str]]:
    return [ self.setup_auth(*auth) for auth in authorizations ]
```

File: sewer/auth.py [hunk 9]
```diff
@@ -1,42 +1,115 @@
...
+        for auth in authorizations:
+            self.setup_auth(*auth)
+
+    def is_ready_cert(self, authorizations: Sequence[Tuple[str, str, str]]):
```
Comment (AlecTroemel) [inline sewer/auth.py:96]: what is you plan for this method, and `is_ready_auth`? I'm not seeing them used in the client

File: sewer/auth.py (hunk 8 above)
Comment (AlecTroemel) [reply sewer/auth.py:92]: after a bit more thought, why not only have clients implement the multiple authorizations methods. something like this..

```python
# previously called setup_cert
def setup(self, authorizations: Sequence[Tuple[str, str, str]]) -> Sequence[Tuple[str, str, str]]:
     "setup list of authorizations - add dns records, files, etc."
     raise NotImplementedError("setup method must be implemented")

# previously called cleanup_cert
def cleanup(self, authorizations: Sequence[Tuple[str, str, str]]):
     "cleanup list of authorizations - removes dns records, files, etc."
     raise NotImplementedError("cleanup method must be implemented")
````

the clients can break that down task however they want (batched together, per auth, etc). One way to look at it would be that this base class just defines a common API for `client.py` to use

Comment (mmaney) [reply sewer/auth.py:92]: 3-in-1 reply!

Using setup_cert is on the todo list - that just requires collecting all the challenges instead of firing them off one by one.  And no, they no longer have to return a filtered list - that was delegating work to the driver that the client code can do in one line!

And yes, the endpoint is that drivers all accept the *_cert calls.  Everything else is for compatibility with existing code (everything in dns_providers).

Oh.  I adopted the setup and clear prefixes (did I change one of those along the way?), and when I was looking for the name suffix I realized that the batches are going to naturally be the challenges for one certificate request, and that sounded better to me than whatever alternatives I tried.

File: sewer/auth.py (hunk 9 above)
Comment (mmaney) [reply sewer/auth.py:96]: This was a late thought.  Use is obvious - after setup_cert, call is_ready_cert and driver MAY verify that the challenge data is in place... this is more an issue for some dns services, and when it's worth checking and how to check are best left to the driver.

Thanks for looking it over.  I'm going to finish this set of changes RealSoonNow... time just keeps slipping away from me. :-(

Comment (AlecTroemel) [reply sewer/auth.py:96]: ah, I see.

exited to see your changes, thanks for the work on this!

File: sewer/client.py (hunk 6 above)
Comment (mmaney) [reply sewer/client.py:253]: #169 is on the fast path to commit.  Great appreciation if you could sanity check it, since it's All Your Fault.  :-)

File: sewer/auth.py (hunk 8 above)
Comment (mmaney) [reply sewer/auth.py:92]: So while looking at the next step in reworking this series as if I'd known what I was doing <wink>, I found out where my choice of prefixes came from.  You gave me _cleanup_, and perhaps because I'd recently been mucking about in some existing tests, _setup_ seemed the best complement to it.

I knew I'd stolen them from somewhere!

**Comment by mmaney** [comment]
This PR is dead - as a PR, certainly much of it has been recycled, not only the logging stuff.

@AlecTroemel, I would greatly value your thoughts on #178.  One part of that which is NOT final (and I may make changes this evening, if I can decide what exactly to change) is the handling of the "errata lists" returned by ProviderBase methods.  And some new tests, and at least one example http-01 provider using the new interface, and doc updates, and...

@komuw I just assume you've got an eye on things.  :-)
